AddMonths doesn't handle last day of February correctly

```
--- FAIL: TestAddMonths (0.00s)
    --- FAIL: TestAddMonths/march_->_february (0.00s)
        duration_test.go:217: AddMonths(2015-03-31 08:00:00 +0200 CEST, -1, Europe/Berlin):
        duration_test.go:218:   want=2015-02-28 08:00
        duration_test.go:219:   have=2015-03-31 09:00
    --- FAIL: TestAddMonths/february_->_march (0.00s)
        duration_test.go:217: AddMonths(2015-02-28 08:00:00 +0100 CET, 1, Europe/Berlin):
        duration_test.go:218:   want=2015-03-31 08:00
        duration_test.go:219:   have=2015-03-28 08:00
    --- FAIL: TestAddMonths/march_->_february_29th_(leap_year) (0.00s)
        duration_test.go:217: AddMonths(2016-03-31 08:00:00 +0200 CEST, -1, Europe/Berlin):
        duration_test.go:218:   want=2016-02-29 08:00
        duration_test.go:219:   have=2016-03-31 09:00
```

--- a/duration.go
+++ b/duration.go
@@ -25,6 +25,7 @@ func addDuration(t0 time.Time, l *time.Location, d time.Duration) time.Time {
 	t0 = t0.In(l)
 	t1 := t0.Add(d)
 
+	// account for changes due to DST
 	_, oldOffset := t0.Zone()
 	_, newOffset := t1.Zone()
 	diff := time.Duration(oldOffset-newOffset) * time.Second
@@ -38,15 +39,25 @@ func addDuration(t0 time.Time, l *time.Location, d time.Duration) time.Time {
 func AddMonths(t time.Time, months int, l *time.Location) time.Time {
 	t = t.In(l)
 	t1 := t.AddDate(0, months, 0)
-
 	d0, d1 := t.Day(), t1.Day()
-	if d1 != d0 && d1 == 1 {
-		// we overshot by 1 day due to normalization (see documentation of AddDate)
-		t1 = t1.Add(-day)
+	if d1 != d0 {
+		// due to normalization in AddDate we might need to correct
+		// t1. See documentation of AddDate.
+		if months > 0 && d1 == 1 {
+			// we overshot by 1 day
+			t1 = t1.Add(-day)
+		}
+		if months < 0 {
+			if t1.Month() == t.Month() {
+				// ended up in the same month due to normalization
+				// worst case: 2015-03-31 - 1month = 2015-02-31 = 2015-03-03
+				diff := time.Duration(t1.Day()) * day
+				t1 = t1.Add(-diff)
+			}
+		}
 	}
 
 	if IsLastDayOfMonth(t) {
-		// we want to stick to the last day of the month
 		_, oldOffset := t1.Zone()
 		tmp, m1 := t1, t1.Month()
 		for {
@@ -73,5 +84,7 @@ func IsLastDayOfMonth(t time.Time) bool {
 		return false
 	}
 
+	// by definition t is the last day of the month if adding another day would
+	// result in a time within the next month.
 	return t.Add(day).Month() != t.Month()
 }

--- a/duration.go
+++ b/duration.go
@@ -22,9 +22,10 @@ func AddWeeks(t time.Time, weeks int, l *time.Location) time.Time {
 }
 
 func addDuration(t0 time.Time, l *time.Location, d time.Duration) time.Time {
-	t1 := t0.Add(d).In(l)
+	t0 = t0.In(l)
+	t1 := t0.Add(d)
 
-	_, oldOffset := t0.In(l).Zone()
+	_, oldOffset := t0.Zone()
 	_, newOffset := t1.Zone()
 	diff := time.Duration(oldOffset-newOffset) * time.Second
 
@@ -35,11 +36,10 @@ func addDuration(t0 time.Time, l *time.Location, d time.Duration) time.Time {
 // the last day of its month then the returned time t' represents the last day
 // of the next month after t. Otherwise t.Day() will equal t'.Day().
 func AddMonths(t time.Time, months int, l *time.Location) time.Time {
-	t1 := t.In(l).AddDate(0, months, 0)
+	t = t.In(l)
+	t1 := t.AddDate(0, months, 0)
 
-	d0 := t.Day()
-	d1 := t1.Day()
-
+	d0, d1 := t.Day(), t1.Day()
 	if d1 != d0 && d1 == 1 {
 		// we overshot by 1 day due to normalization (see documentation of AddDate)
 		t1 = t1.Add(-day)
@@ -47,6 +47,7 @@ func AddMonths(t time.Time, months int, l *time.Location) time.Time {
 
 	if IsLastDayOfMonth(t) {
 		// we want to stick to the last day of the month
+		_, oldOffset := t1.Zone()
 		tmp, m1 := t1, t1.Month()
 		for {
 			tmp = tmp.Add(day)
@@ -56,6 +57,10 @@ func AddMonths(t time.Time, months int, l *time.Location) time.Time {
 			t1 = tmp
 		}
 
+		if _, newOffset := t1.Zone(); newOffset != oldOffset {
+			diff := time.Duration(oldOffset-newOffset) * time.Second
+			t1 = t1.Add(diff)
+		}
 	}
 
 	return t1
@@ -64,7 +69,7 @@ func AddMonths(t time.Time, months int, l *time.Location) time.Time {
 // IsLastDayOfMonth returns true if t is any time within the last day of the
 // month of t.
 func IsLastDayOfMonth(t time.Time) bool {
-	if t.Day() < 29 {
+	if t.Day() < 28 {
 		return false
 	}
 

--- a/duration_test.go
+++ b/duration_test.go
@@ -176,6 +176,36 @@ func TestAddMonths(t *testing.T) {
 			want:   time.Date(2016, 10, 1, 8, 0, 0, 0, berlin),
 			months: -1,
 		},
+		"february -> march": {
+			t:      time.Date(2015, 02, 28, 8, 0, 0, 0, berlin),
+			want:   time.Date(2015, 03, 31, 8, 0, 0, 0, berlin),
+			months: 1,
+		},
+		"march -> february": {
+			t:      time.Date(2015, 03, 31, 8, 0, 0, 0, berlin),
+			want:   time.Date(2015, 02, 28, 8, 0, 0, 0, berlin),
+			months: -1,
+		},
+		"february 28th -> march (leap year)": {
+			t:      time.Date(2016, 02, 28, 8, 0, 0, 0, berlin),
+			want:   time.Date(2016, 03, 28, 8, 0, 0, 0, berlin),
+			months: 1,
+		},
+		"march -> february 28th (leap year)": {
+			t:      time.Date(2016, 03, 28, 8, 0, 0, 0, berlin),
+			want:   time.Date(2016, 02, 28, 8, 0, 0, 0, berlin),
+			months: -1,
+		},
+		"february 29th -> march (leap year)": {
+			t:      time.Date(2016, 02, 29, 8, 0, 0, 0, berlin),
+			want:   time.Date(2016, 03, 31, 8, 0, 0, 0, berlin),
+			months: 1,
+		},
+		"march -> february 29th (leap year)": {
+			t:      time.Date(2016, 03, 31, 8, 0, 0, 0, berlin),
+			want:   time.Date(2016, 02, 29, 8, 0, 0, 0, berlin),
+			months: -1,
+		},
 	}
 
 	for name, c := range cases {

--- a/duration_test.go
+++ b/duration_test.go
@@ -176,35 +176,40 @@ func TestAddMonths(t *testing.T) {
 			want:   time.Date(2016, 10, 1, 8, 0, 0, 0, berlin),
 			months: -1,
 		},
-		"february -> march": {
+		"february → march": {
 			t:      time.Date(2015, 02, 28, 8, 0, 0, 0, berlin),
 			want:   time.Date(2015, 03, 31, 8, 0, 0, 0, berlin),
 			months: 1,
 		},
-		"march -> february": {
+		"march → february": {
 			t:      time.Date(2015, 03, 31, 8, 0, 0, 0, berlin),
 			want:   time.Date(2015, 02, 28, 8, 0, 0, 0, berlin),
 			months: -1,
 		},
-		"february 28th -> march (leap year)": {
+		"february 28th → march (leap year)": {
 			t:      time.Date(2016, 02, 28, 8, 0, 0, 0, berlin),
 			want:   time.Date(2016, 03, 28, 8, 0, 0, 0, berlin),
 			months: 1,
 		},
-		"march -> february 28th (leap year)": {
+		"march → february 28th (leap year)": {
 			t:      time.Date(2016, 03, 28, 8, 0, 0, 0, berlin),
 			want:   time.Date(2016, 02, 28, 8, 0, 0, 0, berlin),
 			months: -1,
 		},
-		"february 29th -> march (leap year)": {
+		"february 29th → march (leap year)": {
 			t:      time.Date(2016, 02, 29, 8, 0, 0, 0, berlin),
 			want:   time.Date(2016, 03, 31, 8, 0, 0, 0, berlin),
 			months: 1,
 		},
-		"march -> february 29th (leap year)": {
+		"march → february 29th (leap year)": {
 			t:      time.Date(2016, 03, 31, 8, 0, 0, 0, berlin),
 			want:   time.Date(2016, 02, 29, 8, 0, 0, 0, berlin),
 			months: -1,
+		},
+		"leap year": {
+			t:      time.Date(2015, 02, 28, 8, 0, 0, 0, berlin),
+			want:   time.Date(2016, 02, 29, 8, 0, 0, 0, berlin),
+			months: 12,
 		},
 	}
 

